### PR TITLE
fix circular dep issue with torch.onnx

### PR DESCRIPTION
**Pull request by anderspapitto**
the problem: if you try to apply the various @symbolic decorators inside most of pytorch, you hit an error due to a circular dependency between the main torch code and torch.onnx.

the fix:

move torch/onnx into torch/onnx/core
introduce a new torch/onnx/init.py, which forwards a couple methods (like _export()), with torch.onnx.core lazily loaded inside of them
move the @symbolic override functionality to torch/onnx/__init.py, and make sure that applying these decorators does not require importing core.

### CODE REVIEW

**Comment by dzhulgakov** [comment]
First of all - thanks for taking this on!

Hm, it doesn't seem to remove a lot of inline imports and instead introduces a few more :(

I didn't dig into details of code dependency - but do you think there's some way of reorganizing stuff to completely get rid of inline imports?

@ezyang @apaszke will probably be better folks to help with advice on untangling this.

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -357,6 +359,7 @@ def _slow_forward(self, *input, **kwargs):
     def __call__(self, *input, **kwargs):
         for hook in self._forward_pre_hooks.values():
             hook(self, input)
+        import torch.jit
```
Comment (apaszke) [inline torch/nn/modules/module.py]: Previously it was ok without those imports. Why are they necessary now?

**Review by apaszke — COMMENTED** [review]
@anderspapitto I'll be happy to help with disentangling this, but I need to see the diagram of dependencies between ONNX (possibly on a function-level) and other PyTorch modules (as they were before this PR). I don't think this is an improvement, because as @dzhulgakov noticed we only ended up having more imports inline.

**Comment by apaszke** [comment]
Actually I think I have an idea for how to centralize onnx in a single place in the codebase. @anderspapitto would you mind having me take a stab at fixing this?

**Comment by apaszke** [comment]
In general I think the approach you took is reasonable. I think we should have a few invariants in place:
1. `torch.onnx.symbolic` should contain **all** symbolic functions, and should be self-contained i.e. have no dependencies on other `torch` modules. This will allow us to import it from wherever we want
2. `torch.onnx.symbolic_override` should also be self-contained (possibly with imports inside). We could move that to `symbolic`.
3. All other functions from our public API should be stubs that call into a lazily-imported submodule (like the `.utils` you created).

I think this would be enough to nicely disentangle concerns. If we want to override/implement a symbolic for a particular function we import `torch.onnx.symbolic` at the top level and get both definitions for both `symbolic_override` and the particular implementation for this op. All PRs that need to change how a given function is exported touch only `torch/onnx/symbolic.py`, and the total overhead of overriding a function is 2 lines.

File: torch/nn/modules/module.py (hunk 1 above)
Comment (anderspapitto) [reply torch/nn/modules/module.py]: I'm not entirely sure. I think it was the case that by the time this runs, something else had caused torch.jit to be imported, and it's no longer the case. I'll dig into it a bit more

File: torch/nn/utils/rnn.py [hunk 2]
```diff
@@ -208,12 +169,36 @@ def pad_packed_sequence(input, *args, **kwargs):
         containing the list of lengths of each sequence in the batch.
 
     """
-    import torch
```
Comment (anderspapitto) [inline torch/nn/utils/rnn.py]: to be clear, achieving the cleanup in this area is effectively the whole point of the diff (to be in line with what Edward proposed in https://github.com/pytorch/pytorch/pull/5420 and https://github.com/pytorch/pytorch/issues/5460)

**Comment by anderspapitto** [comment]
@apaszke all three constraints seem reasonable/correct to me. Since there seems to be some rough consensus that this is worth doing, I'll clean things up so that those points are respected, and so that code churn is minimized

**Comment by dzhulgakov** [comment]
I'm curious whether "torch.onnx.symbolic should contain all symbolic functions, and should be self-contained i.e. have no dependencies on other torch modules" is always possible. What if it requires invoking some utility functions? Alternative is to put symbolic functions next to actual functions declarations. But then we need to make torch.onnx.utils importable everywhere.

**Comment by anderspapitto** [comment]
@dzhulgakov I believe adam addressed that by saying "possibly with imports inside" - i.e. if a particular symbolic implementation needs to see torch.onnx.utils, just put the import inside that implementation

**Comment by apaszke** [comment]
@dzhulgakov Yea, if you *really* need some utility then it should be possible to import it inside, or we can relax the constraint to allow importing `torch/_utils.py`. I think right now, with only minimal code changes, you can get the dependencies down to `_single`, `_pair`, `_triple` from `torch.nn.utils`, but if those are useful there then they should migrate to `torch._utils` anyway.

**Comment by apaszke** [comment]
But in general I think that the rule should be that problems with ONNX imports should only complicate the ONNX part of the codebase and not some unrelated files.

**Comment by anderspapitto** [comment]
had a small git mishap (renamed the branch, github views it as closing the PR)

continued at https://github.com/pytorch/pytorch/pull/5595

**Pull request closed without merging**